jpa-criteria/impl/src/main/java/com/blazebit/persistence/criteria/impl/AbstractBlazeSelectBaseCTECriteria.java

```diff
@@ -398,15 +403,20 @@ public Assignment(SingularAttributePath<?> attributePath, AbstractExpression<?>
     public <X> CriteriaBuilder<X> render(CriteriaBuilder<X> cbs) {
         FullSelectCTECriteriaBuilder<CriteriaBuilder<X>> fullSelectCTECriteriaBuilder = cbs.with(returnType);
         RenderContextImpl context = new RenderContextImpl();
-        render(fullSelectCTECriteriaBuilder, context);
-        return fullSelectCTECriteriaBuilder.end();
+        return renderAndFinalize(fullSelectCTECriteriaBuilder, context);
     }
 
-    protected void render(SelectBaseCTECriteriaBuilder<?> fullSelectCTECriteriaBuilder, RenderContextImpl context) {
+    protected <X> CriteriaBuilder<X> renderAndFinalize(FullSelectCTECriteriaBuilder<CriteriaBuilder<X>> fullSelectCTECriteriaBuilder, RenderContextImpl context) {
+        return ((FullSelectCTECriteriaBuilder<CriteriaBuilder<X>>) render(fullSelectCTECriteriaBuilder, context)).end();
+    }
+
+    protected <CB extends SelectBaseCTECriteriaBuilder<?>> Object render(CB fullSelectCTECriteriaBuilder, RenderContextImpl context) {
```

Comment:
You used to get the same builder type back, but now this can be an instance of i.e. StartOngoingSetOperationCTECriteriaBuilderImpl instead of FullSelectCTECriteriaBuilderImpl. Maintaining a reference to FullSelectCTECriteriaBuilderImpl, it cannot be used to close the StartOngoingSetOperationCTECriteriaBuilderImpl.

Reply:
It's fine as that is an implementation detail. Introducing the necessary super type would lead to type parameter pollution on the use site which would be even worse as that is user facing.